libcudacxx/include/cuda/std/__bit/byteswap.h

```diff
@@ -0,0 +1,173 @@
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _LIBCUDACXX___BIT_BYTESWAP_H
+#define _LIBCUDACXX___BIT_BYTESWAP_H
+
+#include <cuda/std/detail/__config>
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+#include <cuda/std/__bit/has_single_bit.h>
+#include <cuda/std/__concepts/concept_macros.h>
+#include <cuda/std/__type_traits/is_constant_evaluated.h>
+#include <cuda/std/__type_traits/is_integral.h>
+#include <cuda/std/__type_traits/make_unsigned.h>
+#include <cuda/std/climits>
+#include <cuda/std/cstdint>
+
+#if _CCCL_COMPILER(MSVC)
+#  include <intrin.h>
+#endif // _CCCL_COMPILER(MSVC)
+
+_LIBCUDACXX_BEGIN_NAMESPACE_STD
+
+class __byteswap_impl
+{
+  template <class _Half, class _Full>
+  _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI static constexpr _Full __impl_recursive(_Full __val) noexcept
+  {
+    static_assert(sizeof(_Full) == sizeof(_Half) * 2, "Invalid half type passed to __bytswap_impl");
+
+    return static_cast<_Full>(__impl(static_cast<_Half>(__val >> CHAR_BIT * sizeof(_Half))))
+         | (static_cast<_Full>(__impl(static_cast<_Half>(__val))) << CHAR_BIT * sizeof(_Half));
+  }
+
+  _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI static uint32_t __impl_device(uint32_t __val) noexcept
+  {
+    uint32_t __result{};
+    asm("prmt.b32 %0, %1, 0, 0x0123;" : "=r"(__result) : "r"(__val));
+    return __result;
+  }
+
+  _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI static uint64_t __impl_device(uint64_t __val) noexcept
+  {
+    uint32_t __hi{};
+    uint32_t __lo{};
+    asm("mov.b64 {%0, %1}, %2;" : "=r"(__hi), "=r"(__lo) : "l"(__val));
+
+    const uint32_t __swapped_lo = __impl(__hi);
+    const uint32_t __swapped_hi = __impl(__lo);
+    uint64_t __result{};
+    asm("mov.b64 %0, {%1, %2};" : "=l"(__result) : "r"(__swapped_hi), "r"(__swapped_lo));
+
+    return __result;
+  }
+
+public:
+  template <class _Tp>
+  _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI static _CCCL_CONSTEXPR_CXX14 _Tp __impl(_Tp __val) noexcept
+  {
+    _Tp __result{};
+    for (size_t __i{}; __i < sizeof(__val); ++__i)
+    {
+      __result <<= CHAR_BIT;
+      __result |= (__val >> (__i * CHAR_BIT)) & static_cast<_Tp>(UCHAR_MAX);
+    }
+    return __result;
+  }
+
+  _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI static constexpr uint16_t __impl(uint16_t __val) noexcept
+  {
+#if defined(_CCCL_BUILTIN_BSWAP16)
+    return _CCCL_BUILTIN_BSWAP16(__val);
+#  if _CCCL_STD_VER >= 2014
+    if (!__cccl_default_is_constant_evaluated())
+    {
+#    if _CCCL_COMPILER(MSVC)
+      NV_IF_TARGET(NV_IS_HOST, return _byteswap_ushort(__val);)
+#    endif // _CCCL_COMPILER(MSVC)
+    }
+#  endif // _CCCL_STD_VER >= 2014
+#else
+    return (__val << CHAR_BIT) | (__val >> CHAR_BIT);
+#endif
```

Comment:
This does not look correct.

MSVC does not support the builtin, so that conditional should be in the second branch
```suggestion
#if defined(_CCCL_BUILTIN_BSWAP16)
    return _CCCL_BUILTIN_BSWAP16(__val);
#else // ^^^ _CCCL_BUILTIN_BSWAP16 ^^^ / vvv !_CCCL_BUILTIN_BSWAP16 vvv
#  if _CCCL_COMPILER(MSVC)
    if (!_CUDA_VSTD::is_constant_evaluated())
    {
      NV_IF_TARGET(NV_IS_HOST, return _byteswap_ushort(__val);)
    }
#  endif // _CCCL_COMPILER(MSVC)
    return (__val << CHAR_BIT) | (__val >> CHAR_BIT);
#endif // !_CCCL_BUILTIN_BSWAP16
```

Reply:
This might need `!__cccl_default_is_constant_evaluated`